beacon-chain/cache/attestation.go

```diff
@@ -0,0 +1,262 @@
+		forkchoiceAtts: forkchoice.New(),
+	}
+}
+
+func (c *AttestationCache) Add(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.attestations[id]
+	if group == nil {
+		group = &attGroup{
+			slot: att.GetData().Slot,
+		}
+		c.attestations[id] = group
+	}
+
+	if att.IsAggregated() {
+		group.external = append(group.external, att.Clone())
+		return nil
+	}
+
+	local := c.attestations[id].local
+	if local == nil {
+		local = att.Clone()
+	}
+	bit := att.GetAggregationBits().BitIndices()[0]
+	if local.GetAggregationBits().BitAt(uint64(bit)) {
+		return nil
+	}
+	sig, err := aggregateSig(local, att)
+	if err != nil {
+		return errors.Wrapf(err, "could not aggregate signatures")
+	}
+
+	local.GetAggregationBits().SetBitAt(uint64(bit), true)
+	local.SetSignature(sig)
+
+	return nil
+}
+
+func (c *AttestationCache) GetAll() []ethpb.Att {
+	c.RLock()
+	defer c.RUnlock()
+
+	var result []ethpb.Att
+	for _, group := range c.attestations {
+		if group.local != nil {
+			result = append(result, group.local)
+		}
+		for _, a := range group.external {
+			result = append(result, a)
+		}
+	}
+	return result
+}
+
+func (c *AttestationCache) Count() int {
+	c.RLock()
+	defer c.RUnlock()
+
+	count := 0
+	for _, group := range c.attestations {
+		if group.local != nil {
+			count++
+		}
+		count += len(group.external)
+	}
+	return count
+}
+
+func (c *AttestationCache) DeleteCovered(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.attestations[id]
+	if group == nil {
+		return nil
+	}
+
+	local := group.local
+	if local != nil {
+		if covered, err := att.GetAggregationBits().Contains(local.GetAggregationBits()); err != nil {
+			return err
+		} else if covered {
+			group.local = nil
+		}
+	}
+
+	attsToKeep := make([]ethpb.Att, 0, len(group.external))
+	for _, a := range group.external {
+		if covered, err := att.GetAggregationBits().Contains(a.GetAggregationBits()); err != nil {
+			return err
+		} else if !covered {
+			attsToKeep = append(attsToKeep, a)
+		}
+	}
+	group.external = attsToKeep
+
+	if group.local == nil && len(group.external) == 0 {
+		delete(c.attestations, id)
+	}
+
+	return nil
+}
+
+func (c *AttestationCache) PruneBefore(slot primitives.Slot) uint64 {
+	c.Lock()
+	defer c.Unlock()
+
+	var pruneCount uint64
+	for id, group := range c.attestations {
+		if group.slot < slot {
+			if group.local != nil {
+				pruneCount++
+			}
+			pruneCount += uint64(len(group.external))
+			delete(c.attestations, id)
+		}
+	}
+	return pruneCount
+}
+
+func (c *AttestationCache) AggregateIsRedundant(att ethpb.Att) (bool, error) {
```

Comment:
`AggregateIsRedundant` is much more accurate than the current `HasAggregateAttestation`. The current name is misleading because we check if there's an attestation in the pool whose bits cover all bits of the incoming attestation.